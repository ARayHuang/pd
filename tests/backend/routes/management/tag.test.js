const request = require('supertest');
const { connect } = require('ljit-db/mongoose');
const config = require('config');
const mongoose = require('mongoose');
const {
	deleteDocuments,
	insertDocuments,
} = require('../../../../server/fixtures/index');
const { app } = require('../../../../server/management');
const LOGIN = {
	MANAGER: {
		PROVIDER: {
			password: '123qwe',
			username: 'provider01',
		},
	},
};
const TAG_ID = {
	EXISTED: '5ed7351e4f46070bd8e2dfe6',
	NOT_FOUND: '5ed7351e4f46070bd8e2dfe5',
	INVALID: '5ed7351e4f46070bd8e2dfe',
};
let PROVIDER_MANAGER_COOKIE;

beforeAll(async () => {
	try {
		await connect(config.SERVER.MONGO.URL, { debug: false });
		await deleteDocuments();
		await insertDocuments();

		await request(app)
			.post('/api/v1/login')
			.send(LOGIN.MANAGER.PROVIDER)
			.set('Content-Type', 'application/json')
			.set('Accept', 'application/json')
			.expect(201)
			.then(res => {
				PROVIDER_MANAGER_COOKIE = res.headers['set-cookie'];
			});
	} catch (error) {
		throw error;
	}
});

afterAll(() => {
	return mongoose.connection.close();
});

describe('tag', () => {
	describe('get /api/v1/tags/id=:tagId', () => {
		test('200 ok', () => {
			return request(app)
				.get(`/api/v1/tags/id=${TAG_ID.EXISTED}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(200)
				.then(res => {
					expect(res.body).toMatchSnapshot({
						id: expect.any(String),
					});
				});
		});
		test('404 tag not found', () => {
			return request(app)
				.get(`/api/v1/tags/id=${TAG_ID.NOT_FOUND}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(404)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('422 invalid tagId', () => {
			return request(app)
				.get(`/api/v1/tags/id=${TAG_ID.INVALID}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('get /api/v1/tags', () => {
		test('200 ok', async () => {
			let data;

			try {
				await request(app)
					.get('/api/v1/tags')
					.set('Content-Type', 'application/json')
					.set('Accept', 'application/json')
					.set('Cookie', PROVIDER_MANAGER_COOKIE)
					.expect(200)
					.then(res => {
						data = res.body.data;
						expect(res.body).toMatchSnapshot({
							data: expect.any(Array),
						});
					});

				data.forEach(item => {
					expect(item).toMatchSnapshot({
						id: expect.any(String),
					});
				});
			} catch (error) {
				throw error;
			}
		});
		test('422 invalid sort', () => {
			return request(app)
				.get('/api/v1/tags?sort=invalid')
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('post /api/v1/tags', () => {
		test('201 created', () => {
			const body = {
				backgroundColor: '00FF00',
				fontColor: '00FF00',
				status: 'active',
				name: '測試標籤',
			};

			return request(app)
				.post('/api/v1/tags')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(201)
				.then(res => {
					expect(res.body).toMatchSnapshot({
						id: expect.any(String),
					});
				});
		});
		test('409 conflict', () => {
			const body = {
				backgroundColor: '00FF00',
				fontColor: '00FF00',
				status: 'active',
				name: '标签一',
			};

			return request(app)
				.post('/api/v1/tags')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(409)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});

		test('422 invalid backgroundColor', () => {
			const body = {
				backgroundColor: '',
				fontColor: '00FF00',
				status: 'active',
				name: '測試標籤',
			};

			return request(app)
				.post('/api/v1/tags')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('put /api/v1/tags/id=:tagId', () => {
		test('204 no content', () => {
			const body = {
				status: 'disabled',
			};

			return request(app)
				.put(`/api/v1/tags/id=${TAG_ID.EXISTED}`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(204)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('404 tag not found', () => {
			const body = {
				status: 'disabled',
			};

			return request(app)
				.put(`/api/v1/tags/id=${TAG_ID.NOT_FOUND}`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(404)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('422 invalid tagId', () => {
			const body = {
				status: 'disabled',
			};

			return request(app)
				.put(`/api/v1/tags/id=${TAG_ID.INVALID}`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('delete /api/v1/tags/id=:tagId', () => {
		test('204 no content', () => {
			return request(app)
				.delete(`/api/v1/tags/id=${TAG_ID.EXISTED}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(204)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('404 no content', () => {
			return request(app)
				.delete(`/api/v1/tags/id=${TAG_ID.EXISTED}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(404)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('422 invalid tagId', () => {
			return request(app)
				.delete(`/api/v1/tags/id=${TAG_ID.INVALID}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});
});
