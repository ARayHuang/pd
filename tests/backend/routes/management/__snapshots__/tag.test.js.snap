// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tag delete /api/v1/tags/id=:tagId 204 no content 1`] = `Object {}`;

exports[`tag delete /api/v1/tags/id=:tagId 404 no content 1`] = `
Object {
  "code": "888.004.404",
  "message": "找不到标签",
  "type": "NotFoundError",
}
`;

exports[`tag delete /api/v1/tags/id=:tagId 422 invalid tagId 1`] = `
Object {
  "code": "888.004.422",
  "fields": Array [
    "tagId",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`tag get /api/v1/tags 200 ok 1`] = `
Object {
  "data": Any<Array>,
  "numOfItems": 11,
  "numOfPages": 1,
}
`;

exports[`tag get /api/v1/tags 200 ok 2`] = `
Object {
  "backgroundColor": "FBC8D9",
  "fontColor": "CD349D",
  "id": Any<String>,
  "name": "标签一",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 3`] = `
Object {
  "backgroundColor": "FFC1A9",
  "fontColor": "D63E03",
  "id": Any<String>,
  "name": "标签二",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 4`] = `
Object {
  "backgroundColor": "FFDEAD",
  "fontColor": "AC6802",
  "id": Any<String>,
  "name": "标签三",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 5`] = `
Object {
  "backgroundColor": "ADE0F8",
  "fontColor": "0E61B0",
  "id": Any<String>,
  "name": "标签四",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 6`] = `
Object {
  "backgroundColor": "C2E49E",
  "fontColor": "3A6B07",
  "id": Any<String>,
  "name": "标签五",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 7`] = `
Object {
  "backgroundColor": "E0C3FF",
  "fontColor": "6820B5",
  "id": Any<String>,
  "name": "标签六",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 8`] = `
Object {
  "backgroundColor": "97DFCE",
  "fontColor": "018062",
  "id": Any<String>,
  "name": "标签七",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 9`] = `
Object {
  "backgroundColor": "D5C3A2",
  "fontColor": "775310",
  "id": Any<String>,
  "name": "标签八",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 10`] = `
Object {
  "backgroundColor": "D9DE82",
  "fontColor": "858106",
  "id": Any<String>,
  "name": "标签九",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 11`] = `
Object {
  "backgroundColor": "B9C8F9",
  "fontColor": "2636D8",
  "id": Any<String>,
  "name": "标签十",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags 200 ok 12`] = `
Object {
  "backgroundColor": "d4dbe1",
  "fontColor": "000000",
  "id": Any<String>,
  "name": "隐藏标签",
  "status": "disabled",
}
`;

exports[`tag get /api/v1/tags 422 invalid sort 1`] = `
Object {
  "code": "888.004.422",
  "fields": Array [
    "sort",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`tag get /api/v1/tags/id=:tagId 200 ok 1`] = `
Object {
  "backgroundColor": "FBC8D9",
  "fontColor": "CD349D",
  "id": Any<String>,
  "name": "标签一",
  "status": "active",
}
`;

exports[`tag get /api/v1/tags/id=:tagId 404 tag not found 1`] = `
Object {
  "code": "888.004.404",
  "message": "找不到标签",
  "type": "NotFoundError",
}
`;

exports[`tag get /api/v1/tags/id=:tagId 422 invalid tagId 1`] = `
Object {
  "code": "888.004.422",
  "fields": Array [
    "tagId",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`tag post /api/v1/tags 201 created 1`] = `
Object {
  "backgroundColor": "00FF00",
  "fontColor": "00FF00",
  "id": Any<String>,
  "name": "測試標籤",
  "status": "active",
}
`;

exports[`tag post /api/v1/tags 409 conflict 1`] = `
Object {
  "code": "888.004.409",
  "message": "标签已存在",
  "type": "ConflictError",
}
`;

exports[`tag post /api/v1/tags 422 invalid backgroundColor 1`] = `
Object {
  "code": "888.004.422",
  "fields": Array [
    "backgroundColor",
    "backgroundColor",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`tag put /api/v1/tags/id=:tagId 204 no content 1`] = `Object {}`;

exports[`tag put /api/v1/tags/id=:tagId 404 tag not found 1`] = `
Object {
  "code": "888.004.404",
  "message": "找不到标签",
  "type": "NotFoundError",
}
`;

exports[`tag put /api/v1/tags/id=:tagId 422 invalid tagId 1`] = `
Object {
  "code": "888.004.422",
  "fields": Array [
    "tagId",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;
